Create the default PSSession configuration, not tied to a specific PowerShell version.

kalgiz opened the pull request:
## PR Summary
Create the default PSSession configuration, not tied to a specific PowerShell version.

When Enable-PSRemoting command is run, it creates 2 sessions configurations:
- first the same as it was before with the name containing the current version expressed as: 'PowerShell.$PSVersionTable.GitCommitId'
- second with the default name 'PowerShell.6' so that administrators wouldn't have to guess which specific version is installed on the target.

PR addresses the issue: #6470 
## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed - Issue link: https://github.com/PowerShell/PowerShell-Docs/pull/2252
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

kalgiz committed:
[Feature] Create the default PSSession configuration, not tied to a specific Powershell version.

dantraMSFT commented:
Does the new configuration need to be documented?

kalgiz commented:
@dantraMSFT I think it should be. Shall I file an issue for documenting the changes?

kalgiz committed:
[Feature] Tests for Enable-PSRemoting

kalgiz committed:
[Feature] Enable-PSRemoiting command fix.

kalgiz committed:
[Feature] Enable-PSRemoting, dealing with corner case added.

dantraMSFT commented:
@kalgiz Yes, in the PowerShell/PowerShell-Docs repro.  I typically submit a PR for the doc changes when I submit the PR for the code changes. 

I would think the doc change would be in  reference\6\Microsoft.PowerShell.Core\about, possibly in about_Session_Configurations.md.

Let me know if you need any assistance.

kalgiz committed:
[Feature] Remoting tests correction

kalgiz committed:
[Feature] Install-PowerShellRemoting script adds the configuration name not tied to any specific Powershell version.

kalgiz committed:
Merge branch 'master' of https://github.com/kalgiz/PowerShell into remoting-conf-name

kalgiz committed:
[Feature]

kalgiz committed:
[Feature] Correcting tests for non Windows platforms.

kalgiz commented:
@dantraMSFT I created a pull request for documentation: https://github.com/PowerShell/PowerShell-Docs/pull/2252

kalgiz committed:
commit

kalgiz committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell into remoting-conf-name

kalgiz committed:
[Feature] Not overwriting existing session by command Enable-PSRemoting.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Core/PSSessionConfiguration.Tests.ps1:
```diff
@@ -838,6 +838,27 @@ namespace PowershellTestConfigNamespace
         }
     }
+
+    Describe "Validate Enable-PSSession Cmdlet" -Tags @("Feature", 'RequireAdminOnWindows') {
```
There should be a test case where if you run `Enable-PSRemoting` or `Install-PowerShellRemoting.ps1` and either the PowerShell.6 endpoint or PowerShell.6.x.x endpoint exists, we don't overwrite it (modify some aspect of it's configuration like the ACL, run the cmdlet/script, verify the ACL wasn't reset)

kalgiz commented:
I see that here: https://github.com/kalgiz/PowerShell/blob/67a87729f4e16b223a5f92e9ac2e6bb2a2f5152b/src/System.Management.Automation/engine/remoting/commands/CustomShellCommands.cs#L5000

We were changing the pssessions permission for all existing configurations... I am wondering what is the goal of it and if we still want to do so, if we don't want to overwrite the session configurations?

kalgiz commented:
Also what for me is not so easy to understand in the current code is this condition: 
[https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/remoting/commands/CustomShellCommands.cs#L4861]
`$pscmdlet.ShouldProcess($qMessage, $captionForRegisterDefault)`
What does it mean?

kalgiz committed:
[Feature] Enable-PSRemoting command correction.

kalgiz committed:
Merge branch 'master' of https://github.com/kalgiz/PowerShell into remoting-conf-name

mirichmo commented:
Regarding Enable-PSRemoting and Install-PowerShellRemoting, one thing to keep in mind (a reason for the overwrite behavior) is that they can be used to repair damaged endpoints if the permissions or configurations become invalid. Enable-PSRemoting will either turn on and configure remoting or it will reset remoting back to its original state.

You could tweak the behavior of Install-PowerShellRemoting to support -Force if an endpoint already exists. It would technically be a breaking change, but that could be valuable as it would protect against accidental overwrite.

Side Note:
I originally wrote Install-PowerShellRemoting.ps1 because I/we had not yet ported Enable-PSRemoting & friends. The only reason we kept it around afterward was the "Nano Server Scenario." We needed a way to register an endpoint for a different instance of PowerShell (not the version that was running). There is a separate issue ( #4528 ) to formalize that scenario into a real cmdlet, but it hasn't been implemented yet.

kalgiz committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell

PaulHigin commented on src/System.Management.Automation/engine/remoting/commands/CustomShellCommands.cs:
```diff
@@ -4808,6 +4808,101 @@ public sealed class EnablePSRemotingCommand : PSCmdlet
+    # Section 1:
+    # Move pwrshplugin.dll from $PSHOME to the endpoint directory
+    #
+    $pwshVersion = $configurationName.Replace(""PowerShell."", """")
```
It is not clear to me why "PowerShell." is removed from the configuration name.  Can you add examples in the comment section that illustrates the expected behavior, i.e., before string then after string?

kalgiz replied:
I wanted to stick to the original version of the code where the plugin path was: '$env:WINDIR\System32\PowerShell\' + powershell_version (for the tied to the version endpoint it would be $PSVersionTable.PSVersion). To be consistent for our new endpoint the plugin path is: '$env:WINDIR\System32\PowerShell\6'.
I added the comment in code.

PaulHigin commented on src/System.Management.Automation/engine/remoting/commands/CustomShellCommands.cs:
```diff
@@ -4836,70 +4931,15 @@ public sealed class EnablePSRemotingCommand : PSCmdlet
-                set-item -WarningAction SilentlyContinue wsman:\localhost\plugin\{0}\Quotas\MaxShellsPerUser -value ""25"" -confirm:$false
-                set-item -WarningAction SilentlyContinue wsman:\localhost\plugin\{0}\Quotas\MaxIdleTimeoutms -value {4} -confirm:$false
-                restart-service winrm -confirm:$false
+            # Create the default PSSession configuration, not tied to any specific PowerShell version.
```
Again, please add an example name in the comment to illustrate how the endpoint name should look.  I think this will help other developers reading the code.

kalgiz replied:
Comment added.

PaulHigin commented on src/powershell-native/Install-PowerShellRemoting.ps1:
```diff
@@ -103,84 +103,103 @@ function Generate-PluginConfigFile
+        $targetPsVersion = $PSVersionTable.PSVersion.ToString()
+    }
+
+    if ($VersionIndependent) {
```
Please add a comment that illustrates the effect of this, i.e., before and after strings.

kalgiz replied:
Comment added.

PaulHigin reviewed (changes requested):
Very nice refactoring.  This is complex code and it is a bit cleaner now.

kalgiz committed:
Merge branch 'master' of https://github.com/kalgiz/PowerShell

kalgiz committed:
Merge branch 'master' of https://github.com/kalgiz/PowerShell into remoting-conf-name

kalgiz committed:
[Feature] Comments for endpoints' name clarifications added.

kalgiz commented:
Maybe it would be better if we reset the sessions configurations to the original state (providing they exist) only if parameter -Force is on? Right now we are overwriting some permission settings...

PaulHigin reviewed (approved):
LGTM

dantraMSFT commented:
I have a couple of high-level questions:

1: Shouldn't the scripts be calling Set-StrictMode Latest?
2: Shouldn't ErrorActionPreference be set to Stop?
There are a number of places where errors can occur (copy-item for example) and ErrorAction is not used.
3: I assume that an elevated process is needed, correct? Should the script(s) explicitly test for it instead relying on an access denied somewhere in the script?
4: Have you considered running the script analyzer against the scripts?

kalgiz committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell

kalgiz committed:
Merge branch 'master' of https://github.com/kalgiz/PowerShell

kalgiz committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell

kalgiz committed:
Merge branch 'master' of https://github.com/kalgiz/PowerShell into remoting-conf-name

kalgiz committed:
Code fixes.

kalgiz commented on src/powershell-native/Install-PowerShellRemoting.ps1:
```diff
@@ -103,84 +103,106 @@ function Generate-PluginConfigFile
+function Install-PluginEndpoint {
+    param (
+        [switch]
+        # [parameter()]
```
Done.

kalgiz commented on src/System.Management.Automation/engine/remoting/commands/CustomShellCommands.cs:
```diff
@@ -4808,6 +4808,105 @@ public sealed class EnablePSRemotingCommand : PSCmdlet
 }}
 
+function Copy-PluginToEndpoint
+{{
```
I followed the pattern in the file but single brackets in powershell code given as a string format throw exceptions.

PaulHigin replied:
These are needed to escape the '{' character, because the string is being used with .net formatting, i.e., string.Format, via PowerShell's -f keyword.  String.Format uses the '{' character to indicate where formatting arguments are placed.  The code has a comment explaining this.

kalgiz commented on src/System.Management.Automation/engine/remoting/commands/CustomShellCommands.cs:
```diff
@@ -4808,6 +4808,105 @@ public sealed class EnablePSRemotingCommand : PSCmdlet
+    [Parameter()] [string] $endpointDir
+)
+    $resolvedPluginInstallPath = """"
+    $pluginInstallPath = Join-Path ""$env:WINDIR\System32\PowerShell"" $endpointDir
```
Done.

kalgiz commented on src/powershell-native/Install-PowerShellRemoting.ps1:
```diff
@@ -103,84 +103,106 @@ function Generate-PluginConfigFile
+    Write-Verbose "Using PowerShell Version: $targetPsVersion" -Verbose
 
-$pluginBasePath = Join-Path "$env:WINDIR\System32\PowerShell" $targetPsVersion
+    $pluginBasePath = Join-Path "$env:WINDIR\System32\PowerShell" $targetPsVersion
```
Done.

kalgiz commented on src/System.Management.Automation/engine/remoting/commands/CustomShellCommands.cs:
```diff
@@ -4836,70 +4935,16 @@ public sealed class EnablePSRemotingCommand : PSCmdlet
-                restart-service winrm -confirm:$false
+            # Create the default PSSession configuration, not tied to specific PowerShell version
+            # e. g. 'PowerShell.6'.
+            $powershellNr = $PSVersionTable.PSVersion.ToString()
```
Changed to $powershellVersionMajor.

kalgiz committed:
Set-StrictoMode Latest added to the psremoting scripts.

kalgiz committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell

kalgiz committed:
Merge branch 'master' of https://github.com/kalgiz/PowerShell

kalgiz committed:
and ShouldProcess in Intall-PowerShellRemoting script.

kalgiz committed:
Enable-PSRemoting corrections after running SA.

kalgiz committed:
-ErrorAction Stop in Install-PowerShellRemoting module.

kalgiz committed:
-ErrorAction Stop used for commands in Enable-PSRemoting definition.

kalgiz commented:
@dantraMSFT 

1. I added call to Set-StrictMode Latest in both scripts: Install-PowerShellRemoting.ps1 and also CustomShellCommands where Enable-PSRemoting is defined as PowerShell script.
2. In both scripts (the same as above) I added "-ErrorAction Stop" for command for which it seemed reasonable for me. 
3. It seems currently there is a check in Install-PowerShellRemoting.ps1 script for Administrator rights:
`if (! ([Security.Principal.WindowsPrincipal] [Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole] "Administrator"))
{
    Write-Error "WinRM registration requires Administrator rights. To run this cmdlet, start PowerShell with the `"Run as administrator`" option."
    return
}`
4. I added the ScriptAnalyzer against the scripts and tried to address the issues that it indicated. I left Write-Host in Install-PowerShellRemoting.ps1 though and I am wondering if we should get rid of it as it informs user about steps performed.

kalgiz committed:
check for administrator rights moved to the beginning of the script.

kalgiz commented on assets/Product.wxs:
```diff
@@ -6,28 +6,29 @@
   <?define InfoURL="https://github.com/PowerShell/PowerShell" ?>
   <?define ProductName = "$(env.ProductName)" ?>
   <?define ProductCode = "$(env.ProductCode)" ?>
-  <?define ProductSimpleVersionWithNameAndOptionalArchitecture = "$(var.ProductName) $(env.SimpleProductVersion) ($(sys.BUILDARCH))"?>
```
There was some weird merge in my branch, I reset it to the stage before.

kalgiz committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell

kalgiz committed:
Merge branch 'master' of https://github.com/kalgiz/PowerShell into remoting-conf-name

kalgiz committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell

kalgiz committed:
Merge branch 'master' of https://github.com/kalgiz/PowerShell into remoting-conf-name

kalgiz committed:
[Feature] Write-Error if Install-PowerShellRemoting is called without -Force option and the default configurations already exist.

kalgiz commented on src/powershell-native/Install-PowerShellRemoting.ps1:
```diff
@@ -97,90 +104,119 @@ function Generate-PluginConfigFile
+    # If endpoint exists and -Force parameter was not used, the endpoint would not be overwritten.
+    if ($endpoint -and !$Force)
+    {
+        Write-Host "Endpoint $pluginEndpointName already exists." -foregroundcolor Magenta
```
Fixed.

kalgiz committed:
Merge branch 'master' of https://github.com/kalgiz/PowerShell into remoting-conf-name

kalgiz committed:
Merge branch 'remoting-conf-name' of https://github.com/kalgiz/PowerShell into remoting-conf-name

kalgiz committed:
Correcting erorr appearing because of the Strict Mode

kalgiz committed:
[Feature] Write-Error with -Category and -ErrorId parameters.

kalgiz commented on src/powershell-native/Install-PowerShellRemoting.ps1:
```diff
@@ -155,7 +155,7 @@ function Install-PluginEndpoint {
     if ($endpoint -and !$Force)
     {
-        Write-Host "Endpoint $pluginEndpointName already exists." -foregroundcolor Magenta
+        Write-Error "Endpoint $pluginEndpointName already exists."
```
Fixed.

kalgiz committed:
[Feature] Remove check of the value of non assigned variable.

kalgiz committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell

kalgiz committed:
Merge branch 'master' of https://github.com/kalgiz/PowerShell

kalgiz committed:
Merge branch 'master' of https://github.com/kalgiz/PowerShell into remoting-conf-name
